### 作業目的:了解決策樹的節點分支依據
本次作業可參考簡報中的延伸閱讀[訊息增益](https://medium.com/jameslearningnote/%E8%B3%87%E6%96%99%E5%88%86%E6%9E%90-%E6%A9%9F%E5%99%A8%E5%AD%B8%E7%BF%92-%E7%AC%AC3-5%E8%AC%9B-%E6%B1%BA%E7%AD%96%E6%A8%B9-decision-tree-%E4%BB%A5%E5%8F%8A%E9%9A%A8%E6%A9%9F%E6%A3%AE%E6%9E%97-random-forest-%E4%BB%8B%E7%B4%B9-7079b0ddfbda)部分

#### Question:
若你是決策樹，下列兩種分類狀況(a,b)，你會選擇哪種做分類？為什麼？

<img src='hw_1.png' style='width:500px'>

In [2]:
import math

In [8]:
### Answer:
ep = -(1/2 * math.log2(1/2) + 1/2 * math.log2(1/2))
el = -(3/4 * math.log2(3/4) + 1/4 * math.log2(1/4))
er = -(1/4 * math.log2(1/4) + 3/4 * math.log2(3/4))
print(ep, el, er)
ep - 1/2*el - 1/2*er 

1.0 0.8112781244591328 0.8112781244591328


0.1887218755408671

In [9]:
ep = -(1/2 * math.log2(1/2) + 1/2 * math.log2(1/2))
el = -(3/7 * math.log2(3/7) + 4/7 * math.log2(4/7))
er = -(1 * math.log2(1))
print(ep, el, er)
ep - 7/8*el - 1/8*er 

1.0 0.9852281360342515 -0.0


0.13792538097002993

In [11]:

gp = 1 - ((0.5)**2 +(0.5)**2)
gl = 1 -
gr = 1 -
print(gp,gl,gr)
gp-gl-gr

0.5 0 0


0.5

In [18]:
def single_split_gini(p1,p2,l1,l2,r1,r2):
    def get_gini(class1, class2):
        return 1 - ((class1/(class1+class2))**2 + (class2/(class1+class2))**2)
    ratio = (l1+l2)/(p1+p2)
    gp = get_gini(p1,p2)
    gl = get_gini(l1,l2)
    gr = get_gini(r1,r2)
    print(f'original gini: {round(gp,3)},splited_left: {round(gl,3)},splited_right: {round(gr,3)}, result is {round(gp-ratio*gl-(1-ratio)*gr,3)}')


In [19]:
single_split_gini(20,20,15,20,5,0)

original gini: 0.5,splited_left: 0.49,splited_right: 0.0, result is 0.071


In [20]:
single_split_gini(20,20,15,5,5,15)

original gini: 0.5,splited_left: 0.375,splited_right: 0.375, result is 0.125


依據方案a, 會得到0.189的information gain,依據方案b, 會得到0.138的information gain
故選用方案a會是比較好的

### 閱讀作業

決策樹根據計算分割準則的不同(ex: Entropy, Gini, Gain ratio)，可分為ID3, C4.5, CART樹的算法，請同學閱讀下列文章，來更加了解決策樹的算法。

[決策樹(ID3, C4.5, CART)](https://blog.csdn.net/u010089444/article/details/53241218)

In [ ]:
閱畢，CART需要離散取值,
將連續型的資料，用值來當閥值，大於或小於的分屬於兩個類別